graylog2-server/src/test/java/org/graylog/security/certutil/CertRenewalServiceTest.java

```diff
@@ -0,0 +1,94 @@
+import org.graylog.scheduler.clock.JobSchedulerClock;
+import org.graylog2.plugin.certificates.RenewalPolicy;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.junit.jupiter.api.Test;
+
+import java.math.BigInteger;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+
+import static org.graylog.security.certutil.CertConstants.KEY_GENERATION_ALGORITHM;
+import static org.graylog.security.certutil.CertConstants.SIGNING_ALGORITHM;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class CertRenewalServiceTest {
+    private X509Certificate generate(final DateTime now, final int subtractMinutes, final int addMinutes) throws OperatorCreationException, CertificateException, NoSuchAlgorithmException {
+        final Date notBefore = now.minusMinutes(subtractMinutes).toDate();
+        final Date notAfter = now.plusMinutes(addMinutes).toDate();
+
+        KeyPairGenerator keyGen = KeyPairGenerator.getInstance(KEY_GENERATION_ALGORITHM);
+        java.security.KeyPair keyPair = keyGen.generateKeyPair();
+
+        final ContentSigner contentSigner = new JcaContentSignerBuilder(SIGNING_ALGORITHM).build(keyPair.getPrivate());
+        final X500Name x500Name = new X500Name("CN=graylog.test");
+        final X509v3CertificateBuilder certificateBuilder =
+                new JcaX509v3CertificateBuilder(x500Name,
+                        BigInteger.valueOf(now.getMillis()),
+                        notBefore,
+                        notAfter,
+                        x500Name,
+                        keyPair.getPublic());
+
+        return new JcaX509CertificateConverter()
+                .setProvider(new BouncyCastleProvider()).getCertificate(certificateBuilder.build(contentSigner));
+    }
+
+    @Test
+    public void testCertRenewalCalculation() throws CertificateException, NoSuchAlgorithmException, OperatorCreationException {
+        final var now = DateTime.now(DateTimeZone.UTC);
+        final JobSchedulerClock clock = new JobSchedulerTestClock(now);
+        final var nextRun = clock.nowUTC().plusMinutes(30);
+        final var service = new CertRenewalServiceImpl(clock);
+        // 2 hours is our smalles interval in the FE, so I'm testing only small intervals and expect larger intervals to work accordingly
+        // with a threshold of 10%, a cert should either be invalid if it's no longer valid in 12min (10% of 120min) or if it get's invalid until the next run of the cert checker
+        final var policy = new RenewalPolicy(RenewalPolicy.Mode.MANUAL, "PT2H");
+
+        // renewal check every 30min, so the following should be still valid
+        final var cert1 = generate(now, 0, 35);
+        assertFalse(service.needsRenewal(nextRun, policy, cert1));
```

Comment:
The test suggests that the service interface should return a testable value. Then we wouldn't need to test a package private method and could use a proper return type of the interface.